File: APP03022-mrch.backend.somx.catalogos-api/src/main/java/com/sodimac/catman/api/controller/ConversionController.java
package com.sodimac.catman.api.controller;

import com.sodimac.catman.api.model.dto.*;
import com.sodimac.catman.api.service.ConversionService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/conversions")
@Tag(name = "Conversiones", description = "API para gestión de conversiones entre elementos")
public class ConversionController {

    private final ConversionService svc;
    public ConversionController(ConversionService svc) { this.svc = svc; }

    @Operation(summary = "Buscar conversiones con filtros y paginación")
    @GetMapping
    public ResponseEntity<ConversionPageResponse> search(
            @RequestParam(required = false) Integer idElementoOrigen,
            @RequestParam(required = false) Integer idElemento,
            @RequestParam(required = false) String elemento,
            @RequestParam(required = false) String valorElemento,
            @RequestParam(required = false) String catalogoOrigen,
            @RequestParam(required = false) Integer estatus,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int pageSize,
            @RequestParam(defaultValue = "createdAt") String sortBy,
            @RequestParam(defaultValue = "desc") String sortDir) {

        Sort sort = sortDir.equalsIgnoreCase("asc") ? Sort.by(sortBy).ascending() : Sort.by(sortBy).descending();
        Pageable pageable = PageRequest.of(Math.max(0, page - 1), pageSize, sort);
        return ResponseEntity.ok(svc.search(idElementoOrigen, idElemento, elemento, valorElemento, catalogoOrigen, estatus, pageable));
    }

    @Operation(summary = "Consultar conversión por ID")
    @GetMapping("/{id}")
    public ResponseEntity<ConversionDto> getById(@PathVariable Integer id) {
        return ResponseEntity.ok(svc.getById(id));
    }

    @Operation(summary = "Crear conversión")
    @PostMapping
    public ResponseEntity<ConversionDto> create(
            @RequestHeader(value = "X-User-Id", defaultValue = "system") String userId,
            @Valid @RequestBody ConversionCreateDto dto) {
        return ResponseEntity.status(HttpStatus.CREATED).body(svc.create(dto, userId));
    }

    @Operation(summary = "Editar conversión")
    @PutMapping("/{id}")
    public ResponseEntity<ConversionDto> update(
            @PathVariable Integer id,
            @RequestHeader(value = "X-User-Id", defaultValue = "system") String userId,
            @Valid @RequestBody ConversionUpdateDto dto) {
        return ResponseEntity.ok(svc.update(id, dto, userId));
    }

    @Operation(summary = "Cambiar conversión principal")
    @PatchMapping("/{id}/principal")
    public ResponseEntity<ConversionDto> setPrincipal(
            @PathVariable Integer id,
            @RequestHeader(value = "X-User-Id", defaultValue = "system") String userId,
            @RequestBody Map<String, Boolean> body) {
        Boolean isPrincipal = body.getOrDefault("esPrincipal", body.getOrDefault("conversionPrincipal", true));
        return ResponseEntity.ok(svc.setPrincipal(id, isPrincipal, userId));
    }

    @Operation(summary = "Eliminar conversión individual")
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(
            @PathVariable Integer id,
            @RequestHeader(value = "X-User-Id", defaultValue = "system") String userId) {
        svc.delete(id);
        return ResponseEntity.noContent().build();
    }

    @Operation(summary = "Eliminar múltiples conversiones")
    @DeleteMapping
    public ResponseEntity<Void> deleteMultiple(
            @RequestHeader(value = "X-User-Id", defaultValue = "system") String userId,
            @RequestBody Map<String, List<Integer>> body) {
        List<Integer> ids = body.get("ids");
        if (ids != null && !ids.isEmpty()) svc.deleteMultiple(ids);
        return ResponseEntity.noContent().build();
    }
}

File: APP03022-mrch.backend.somx.util-api/src/controllers/conversion.controller.ts
import type { Request, Response, NextFunction } from 'express';
import * as conversionService from '@/services/conversion.service.js';
import { ConversionCreateSchema, ConversionUpdateSchema } from '@/dto/conversion.dto.js';

function userId(req: Request): string {
    const h = req.header('X-User-Id');
    return h && h.trim() !== '' ? h : 'system';
}

function parseSortDir(raw: unknown): 'ASC' | 'DESC' {
    const s = String(raw ?? 'desc').toLowerCase();
    return s === 'asc' ? 'ASC' : 'DESC';
}

function toIntOrNull(v: unknown): number | null {
    if (v === undefined || v === null || v === '') return null;
    const n = Number(v);
    return Number.isFinite(n) ? n : null;
}

function toStringOrNull(v: unknown): string | null {
    if (v === undefined || v === null) return null;
    const s = String(v);
    return s.trim() === '' ? null : s;
}

export async function search(req: Request, res: Response, next: NextFunction) {
    try {
        const q = req.query;
        const result = await conversionService.search({
            sourceElementId: toIntOrNull(q.idElementoOrigen),
            targetElementId: toIntOrNull(q.idElemento),
            elemento: toStringOrNull(q.elemento),
            valor: toStringOrNull(q.valorElemento),
            catalogoOrigen: toStringOrNull(q.catalogoOrigen),
            estatus: toIntOrNull(q.estatus),
            page: Math.max(1, Number(q.page ?? 1) || 1),
            pageSize: Math.max(1, Number(q.pageSize ?? 10) || 10),
            sortBy: String(q.sortBy ?? 'createdAt'),
            sortDir: parseSortDir(q.sortDir)
        });
        res.json(result);
    } catch (err) {
        next(err);
    }
}

export async function getById(req: Request, res: Response, next: NextFunction) {
    try {
        const id = Number(req.params.id);
        const result = await conversionService.getById(id);
        res.json(result);
    } catch (err) {
        next(err);
    }
}

export async function create(req: Request, res: Response, next: NextFunction) {
    try {
        const dto = ConversionCreateSchema.parse(req.body);
        const created = await conversionService.create(dto, userId(req));
        res.status(201).json(created);
    } catch (err) {
        next(err);
    }
}

export async function update(req: Request, res: Response, next: NextFunction) {
    try {
        const id = Number(req.params.id);
        const dto = ConversionUpdateSchema.parse(req.body);
        const updated = await conversionService.update(id, dto, userId(req));
        res.json(updated);
    } catch (err) {
        next(err);
    }
}

export async function setPrincipal(req: Request, res: Response, next: NextFunction) {
    try {
        const id = Number(req.params.id);
        const body = (req.body ?? {}) as Record<string, unknown>;
        const isPrincipalRaw = body.esPrincipal ?? body.conversionPrincipal ?? true;
        const isPrincipal = isPrincipalRaw === true || isPrincipalRaw === 'true';
        const updated = await conversionService.setPrincipal(id, isPrincipal, userId(req));
        res.json(updated);
    } catch (err) {
        next(err);
    }
}

export async function deleteOne(req: Request, res: Response, next: NextFunction) {
    try {
        const id = Number(req.params.id);
        await conversionService.deleteById(id);
        res.status(204).send();
    } catch (err) {
        next(err);
    }
}

export async function deleteMultiple(req: Request, res: Response, next: NextFunction) {
    try {
        const body = (req.body ?? {}) as Record<string, unknown>;
        const raw = body.ids;
        if (Array.isArray(raw) && raw.length > 0) {
            const ids = raw.map(v => Number(v)).filter(n => Number.isFinite(n));
            await conversionService.deleteMultiple(ids);
        }
        res.status(204).send();
    } catch (err) {
        next(err);
    }
}

